core/network/src/protocol.rs

```diff
@@ -264,33 +280,74 @@ impl<B: BlockT, S: NetworkSpecialization<B>, H: ExHashT> Protocol<B, S, H> {
 			}
 		};
 
+		fn validate_response<B: BlockT, H: ExHashT, R, F, G>(
+			io: &mut SyncIo,
+			who: NodeIndex,
+			context: &ContextData<B, H>,
+			response_id: message::RequestId,
+			request_visitor: F,
+			request_id: G,
+		) -> Option<R>
+		where F: Fn(&mut Peer<B, H>) -> (&mut Option<time::Instant>, &mut Option<R>),
+			  G: Fn(&R) -> message::RequestId,
+		{
+			let request = {
+				let mut peers = context.peers.write();
+				if let Some(ref mut peer) = peers.get_mut(&who) {
+					let (timestamp, request) = request_visitor(peer);
+
+					*timestamp = None;
+					match mem::replace(request, None) {
```

Comment:
```suggestion
					match request.take() {
```

This might make the `mem` import unneeded